elastic/logstash
# For custom java plugins, set the platform = 'java'

## Release notes
Properly set the platform of custom Java plugins by setting the platform to Java in the generated gemspec.

## What does this PR do?
Includes the `s.platform = 'java'` in the generated gemspec files that are created by the `generateRubySupportFilesForPlugin` function in the rubyUtils.gradle file.

## Why is it important/What is the impact to the user?

The installation and execution of Java custom plugins will not follow the optimal path if they are not marked as for the Java platform.

Background information provided to me by @jsvd 

> Logstash plugins will usually be marked with “platform” as “java” which influences the final gem name:
> ```sh
> ❯ grep -r platform logstash-input-http-3.9.1-java
> logstash-input-http-3.9.1-java/logstash-input-http.gemspec:  s.platform = "java"
> ```
> ```sh
> /tmp ❯ gem fetch logstash-input-http
> Fetching logstash-input-http-3.9.1-java.gem
> Downloaded logstash-input-http-3.9.1-java
> ```

> This [`platform`] is a ruby gem level setting. Ruby gems can either have only ruby code, or ruby code + platform-specific code. If it’s ruby code only then it won’t have a platform setting. If it has platform specific code, by default an empty platform means it’s to be installed on C Ruby and the code is compiled using gcc or similar. If it has java code then the platform is set as “java”
>
> This way the ruby implementation can pick up gems that are for its platform
>
> In the case of Logstash, the JRuby implementation will by default try to install the `-java` version of a gem if it exists, and otherwise try the one with an empty platform
>
> The `s.metadata = { 'logstash_plugin' => 'true', 'logstash_group' => '" + pluginInfo.pluginType + "', 'java_plugin' => 'true'}` is purely metadata gem authors can put there for whatever reason. in our case since we leverage the rubygem “package” manager to install plugins, we use this metadata in the Logstash plugin manager.

> A ruby gem that contains jars should be marked as `s.platform = java`, which really only makes it so that you can’t install it in a normal ruby.
>
> It’s never automatically marked as such, is the author that sets it in the .gemspec (gem manifest).


## Checklist


## Author's Checklist

- [ ] Reviewer: Is there a tag that I need to have Logstash auto-backport this fix?

## How to test this PR locally

1. Check out a plugin, such as  [logstash-input-java_filter_example](https://github.com/logstash-plugins/logstash-filter-java_filter_example)
2. Assemble it using `./gradlew gem`
3. Check the `<name>.gemspec` to confirm the platform is set to Java
4. Check the `<name>.gem` file to confirm it is in the format of `logstash-<type>-<name>-<version>-java.gem`

## Related issues

- Relates to #16626 
- Relates to https://github.com/elastic/logstash/commit/0de53939e952d0ab314c72eaccc083ee4bb465b2
- Relates to https://github.com/elastic/logstash/pull/10620

## Use cases

```
Feature: Plugin build
  Scenario: Custom Java plugin
    When `generateRubySupportFiles` task is ran
    Then `.gemspec` file contains `s.platform = 'java'`
```

## Screenshots

N/A

## Logs

N/A
